### DISCUSSION TITLE
Is it possible to make a peridynamics ADKernel?

### DISCUSSION THREAD

**Discussion by andrisfreimanis** [idaholab/moose#28465] [Q&A Modules: General]
Dear All,
I want to modify peridynamics kernel, but writing the full jacobian for it would be a headache due to nonlocal interaction. Therefore, using an ADKernel seems like an attractive option.

The existing PD implementation inherits from the regular kernel and overrides the computeResidual() method rather than using computeQpResidual(). However, the computeResidual() in Kernel.h is virtual but it is not in ADKernel.h, and if my understanding is correct it is not meant to be overridden. 

But it is better to ask before starting to write a long jacobian. So is it possible to implement a peridynamics ADKernel?

**Comment by GiudGiud** [comment]
Hello

Whether for ADKernel or Kernel you want to override computeQpResidual. Overriding computeResidual is for expert users. 

It should be possible to write an ADKernel for PD. I think some complexity might come from how to handle the bonds breaking but a similar treatment as for non AD might be possible 

AD was added to moose after peridynamics which is why there is no AD in PD right now

**Reply by GiudGiud** [reply]
Note that you can mix and match ADKernel(s) and Kernel(s), so you dont need to implement every PD kernel with AD immediately.
The only issue is material properties, which can be either AD or non AD but not both. They can be converted with this material
https://mooseframework.inl.gov/source/materials/MaterialADConverter.html
but derivatives are generally lost in the process
